### PR TITLE
Unlikelihood token loss implementation

### PR DESCRIPTION
**Pull request by funboarder13920**
- Implement unlikelihood loss from [Neural Text Generation with Unlikelihood Training](https://arxiv.org/abs/1908.04319)
- Tests for unlikelihood loss
- Class refactoring for LossComputeBase
- Fix ppl in statistics that used to be based on the loss which could be different from NLLLoss

### CODE REVIEW

File: .github/workflows/push.yml [hunk 1]
```diff
@@ -148,6 +148,38 @@ jobs:
...
+            -encoder_type transformer_lm \
+            -decoder_type transformer_lm \
+            -src_vocab_size 1000 \
+            -tgt_vocab_size 1000 \
+            -label_smoothing 0.1 \
+            -dec_layers 2 -batch_size 10 \
+            -heads 4 -transformer_ff 64 \
+            -word_vec_size 16 -report_every 5 \
+            -rnn_size 16 -train_steps 10
+    - name: Test LM training with unlieklihood loss
```
Comment (francoishernandez) [inline .github/workflows/push.yml:167]: Typo 'unlikelihood'

File: onmt/utils/loss.py [hunk 2]
```diff
@@ -35,14 +36,21 @@ def build_loss_compute(model, tgt_field, opt, train=True):
...
             unk_index=unk_idx, ignore_index=padding_idx
         )
-    elif opt.label_smoothing > 0 and train:
+    elif opt.label_smoothing > 0:
         criterion = LabelSmoothingLoss(
-            opt.label_smoothing, len(tgt_field.vocab), ignore_index=padding_idx
+            opt.label_smoothing, len(tgt_field.vocab),
+            ignore_index=padding_idx
+        )
+    elif opt.unlikelihood_coeff > 0:
```
Comment (francoishernandez) [inline onmt/utils/loss.py:44]: Shall we assert in the opts validation that `unlikelihood_coeff` isn't compatible with `label_smoothing`?

File: onmt/modules/copy_generator.py [hunk 3]
```diff
@@ -177,7 +177,7 @@ def forward(self, scores, align, target):
         return loss
 
 
-class CommonCopyGeneratorLossCompute(CommonLossCompute):
+class CommonCopyGeneratorLossCompute(LossComputeBase):
```
Comment (francoishernandez) [inline onmt/modules/copy_generator.py:180]: I'm not sure to grasp the whole rationale behind the `CommonLossCompute`/`LossComputeBase` refactoring. Is the last big remaining difference only the log_ppl computation?

Comment (francoishernandez) [reply onmt/modules/copy_generator.py:180]: (Underlying question is: do we really need both `CommonLossCompute` and `LossComputeBase` anymore?)

File: onmt/utils/loss.py [hunk 4]
```diff
@@ -241,16 +312,97 @@ def __init__(self, label_smoothing, tgt_vocab_size, ignore_index=-100):
...
+        return super(ExpandedNLLLoss, self).forward(input, gtruth)
+
+
+class UnlikelihoodTokenLoss(nn.Module):
+    """
+    Unlikelihood token level loss
+    """
+
+    def __init__(self, unlikelihood_coeff, ignore_index=-100):
+        assert 0.0 < unlikelihood_coeff
```
Comment (francoishernandez) [inline onmt/utils/loss.py:338]: Maybe add an explicit message here?

File: onmt/utils/loss.py [hunk 5]
```diff
@@ -241,16 +312,97 @@ def __init__(self, label_smoothing, tgt_vocab_size, ignore_index=-100):
...
+            ignore_index=ignore_index, reduction="none"
+        )
+        self.unlikelihood_coeff = unlikelihood_coeff
+
+    def compute_forbidden_tokens_per_timestep(self, target):
+        expanded_target = target.unsqueeze(-1).expand(
+            target.size(0), target.size(1), target.size(0)
+        ).permute(1, 2, 0)
+
+        ctx_cands = (
```
Comment (francoishernandez) [inline onmt/utils/loss.py:352]: More explicit variable name? Or at least a comment?

**Review by francoishernandez — COMMENTED** [review]
A few comments.

File: onmt/utils/loss.py (hunk 2 above)
Comment (funboarder13920) [reply onmt/utils/loss.py:44]: I can make them mutually exclusive with the parser or I can make both of them compatible at the same time as unlikelihood_coeff can be added to any loss (but label smoothing and unlikelihood are contradictory)

Comment (francoishernandez) [reply onmt/utils/loss.py:44]: Yes, let's go for the mutually exclusive way.

File: onmt/modules/copy_generator.py (hunk 3 above)
Comment (funboarder13920) [reply onmt/modules/copy_generator.py:180]: The _compute_loss, _make_shard_state (I need to move it back to the CommonLossCompute) and the way to use the generator are different between CopyGeneratorLoss and the other classes

Comment (funboarder13920) [reply onmt/modules/copy_generator.py:180]: We can do it in one class, the code is already not very clear, it's not going to be worse. If we do that CopyGenerator will override _compute_loss, _compute_log_ppl and _compute_alignement_loss will only be used in the compute_loss of the main class

Comment (francoishernandez) [reply onmt/modules/copy_generator.py:180]: > We can do it in one class, the code is already not very clear, it's not going to be worse. If we do that CopyGenerator will override _compute_loss, _compute_log_ppl and _compute_alignement_loss will only be used in the compute_loss of the main class

Yes I think this might be a bit better to explicitly override this method instead of having a full class that we don't really know what it's for unless we look at this specific CopyGeneratorLoss.

Comment (funboarder13920) [reply onmt/modules/copy_generator.py:180]: I merged it, the ppl part is not nice. Also there is a normalization args that was not used anywhere, I will investigate to see if the normalization process disappeared by mistake

Comment (funboarder13920) [reply onmt/modules/copy_generator.py:180]: normalization was already not used a year ago https://github.com/OpenNMT/OpenNMT-py/blob/78351302b0761178581d92612b528f6eea529618/onmt/utils/loss.py

**Comment by vince62s** [comment]
@funboarder13920 @francoishernandez would it be worth updating wrt the v3 and merging or shall we drop ?
